transport-native-unix-common-tests/src/main/java/io/netty/channel/unix/tests/UnixChannelUtilTest.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2017 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.channel.unix.tests;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.CompositeByteBuf;
+import io.netty.buffer.UnpooledByteBufAllocator;
+import org.junit.Test;
+
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+
+import static io.netty.channel.unix.Limits.IOV_MAX;
+import static io.netty.channel.unix.UnixChannelUtil.isBufferCopyNeededForWrite;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public class UnixChannelUtilTest {
+    @Test
+    public void testIsBufferCopyNeededForWrite() {
+        ByteBufAllocator alloc = new UnpooledByteBufAllocator(false);
+
+        ByteBuf byteBuf = alloc.directBuffer();
+        assertFalse(isBufferCopyNeededForWrite(byteBuf));
+        assertTrue(byteBuf.release());
+
+        byteBuf = alloc.heapBuffer();
+        assertTrue(isBufferCopyNeededForWrite(byteBuf));
+        assertTrue(byteBuf.release());
+
+        byteBuf = genCompositeByteBuf(alloc, 2, 0);
+        assertFalse(isBufferCopyNeededForWrite(byteBuf));
+        assertTrue(byteBuf.release());
+
+        byteBuf = genCompositeByteBuf(alloc, IOV_MAX + 1, 0);
+        assertTrue(isBufferCopyNeededForWrite(byteBuf));
+        assertTrue(byteBuf.release());
+
+        byteBuf = genCompositeByteBuf(alloc, 0, 2);
+        assertTrue(isBufferCopyNeededForWrite(byteBuf));
+        assertTrue(byteBuf.release());
+
+        byteBuf = genCompositeByteBuf(alloc, 1, 1);
+        assertTrue(isBufferCopyNeededForWrite(byteBuf));
+        assertTrue(byteBuf.release());
+    }
+
+    private static CompositeByteBuf genCompositeByteBuf(ByteBufAllocator alloc, int numDirect, int numHeap) {
+        CompositeByteBuf comp = alloc.compositeBuffer(numDirect + numHeap);
+        List<ByteBuf> byteBufs = new LinkedList<ByteBuf>();
+
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) There is an apparent infinite loop in io.netty.channel.unix.tests.UnixChannelUtilTest.genCompositeByteBuf(ByteBufAllocator, int, int) [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=findbugs%3AIL_INFINITE_LOOP)